pkg/cmd/experimental/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,247 @@
+		return false, fmt.Errorf("No requested diagnostics available"), 0, 1
+	} else if len(common) < len(o.RequestedDiagnostics) {
+		errors = append(errors, fmt.Errorf("Not all requested diagnostics are available"))
+		o.Logger.Errort("notAllReqDiag", `
+Of the requested diagnostics:
+    {{.requested}}
+only these are available:
+    {{.common}}
+The list of all possible is:
+    {{.available}}
+		`, log.Hash{"requested": o.RequestedDiagnostics, "common": common.List(), "available": AvailableDiagnostics.List()})
+	}
+
+	func() { // don't trust discovery/build of diagnostics; wrap panic nicely in case of developer error
+		defer func() {
+			if r := recover(); r != nil {
+				failed = true
+				errors = append(errors, fmt.Errorf("While building the diagnostics, a panic was encountered.\nThis is a bug in diagnostics. Stack trace follows : \n%v", r))
+			}
+		}()
+		detected, detectWarnings, detectErrors := o.detectClientConfig() // may log and return problems
+		for _, warn := range detectWarnings {
+			warnings = append(warnings, warn)
+		}
+		for _, err := range detectErrors {
+			errors = append(errors, err)
+		}
+		if !detected { // there just plain isn't any client config file available
+			o.Logger.Notice("discNoClientConf", "No client configuration specified; skipping client and cluster diagnostics.")
+		} else if rawConfig, err := o.buildRawConfig(); rawConfig == nil { // client config is totally broken - won't parse etc (problems may have been detected and logged)
+			o.Logger.Errorf("discBrokenClientConf", "Client configuration failed to load; skipping client and cluster diagnostics due to error: {{.error}}", log.Hash{"error": err.Error()})
+			errors = append(errors, err)
+		} else {
+			if err != nil { // error encountered, proceed with caution
+				o.Logger.Errorf("discClientConfErr", "Client configuration loading encountered an error, but proceeding anyway. Error was:\n{{.error}}", log.Hash{"error": err.Error()})
+				errors = append(errors, err)
+			}
+			if clientDiags, ok, err := o.buildClientDiagnostics(rawConfig); ok {
+				diagnostics["client"] = clientDiags
+			} else if err != nil {
+				failed = true
+				errors = append(errors, err)
+			}
+
+			if clusterDiags, ok, err := o.buildClusterDiagnostics(rawConfig); ok {
+				diagnostics["cluster"] = clusterDiags
+			} else if err != nil {
+				failed = true
+				errors = append(errors, err)
+			}
+		}
+
+		if hostDiags, ok, err := o.buildHostDiagnostics(); ok {
+			diagnostics["host"] = hostDiags
+		} else if err != nil {
+			failed = true
+			errors = append(errors, err)
+		}
+	}()
+
+	if failed {
+		return failed, kutilerrors.NewAggregate(errors), len(warnings), len(errors)
+	}
+
+	failed, err, numWarnings, numErrors := o.Run(diagnostics)
+	numWarnings += len(warnings)
+	numErrors += len(errors)
+	return failed, err, numWarnings, numErrors
+}
+
+func (o DiagnosticsOptions) Run(diagnostics map[string][]types.Diagnostic) (bool, error, int, int) {
```

Comment:
On Wed, Aug 19, 2015 at 8:28 AM, Luke Meyer <lmeyer@redhat.com> wrote:
>
> I changed it to a container object.
>

And then just removed the whole concept of diagnostic areas... it wasn't
really adding much. Easy to add later if needed.